Add WASM UDF implementation in Rust

wmitros opened the pull request:
This series adds the implementation and usage of rust wasmtime bindings.

The WASM UDFs introduced by this patch are interruptable and use memory allocated using the seastar allocator.

This series includes #11102 (the first two commits) because #11102 required disabling wasm UDFs completely. This patch disables them in the middle of the series, and enables them again at the end.
After this patch, `libwasmtime.a` can be removed from the toolchain. 
This patch also removes the workaround for #https://github.com/scylladb/scylladb/issues/9387 but it hasn't been tested with ARM yet - if the ARM test causes issues I'll revert this part of the change.

wmitros commented:
During the development of this patch I realized an issue that isn't linked to this patch, but to Rust as a whole - the behavior on OOM errors.
Stable Rust `aborts()` on such an error and does not provide methods that allow writing code with a fallible allocations - there are `try_reserve` methods for some Collections but this method is missing for e.g. the `Box` type, which is crucial for us. This could at least partially be helped by `nightly` Rust - the `Box::try_new` is available there, but I would still need to check if the remaining code can be made safe then

wmitros commented on configure.py:
```diff
@@ -1893,7 +1892,7 @@ def configure_abseil(build_dir, mode, mode_config):
               description = TEST {mode}
             rule rust_lib.{mode}
-              command = CARGO_HOME=build/{mode}/rust/.cargo cargo build --release --manifest-path=rust/Cargo.toml --target-dir=build/{mode}/rust -p ${{pkg}}
+              command = CARGO_HOME=build/{mode}/.cargo RUSTFLAGS="--cfg=rustix_use_libc" cargo build --manifest-path=rust/Cargo.toml --target-dir=build/{mode} --profile=rust-{mode}
```
This `RUSTFLAGS` value fixes a weird issue where `rustix` - a rust crate that wasmtime depends on - returns 0 as the system page size when using `linux_raw` instead of `libc` as the backend for systemcalls

sunfishcode replied:
In case it helps, I believe this problem is fixed by https://github.com/bytecodealliance/rustix/pull/385, which is included in rustix 0.35.9.

wmitros replied:
Thanks for the info. I updated the dependencies and confirmed that everything works now without this flag

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/1887/

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/1899/

psarna commented on rust/Cargo.toml:
```diff
@@ -9,3 +9,30 @@ inc = { path = "inc", version = "0.1.0" }
 [lib]
 crate-type = ["staticlib"]
+
+[profile.rust-dev]
```
Nitpick: maybe the prefix should indicate that this is scylla-specific - `scylla-dev` instead of `rust-dev`? `rust-` might suggest that it's some kind of an official standard, and it isn't.

wmitros replied:
`rust-` has a benefit that the directory that contains the build files also has `rust` in its name (currently it's `$builddir/{mode}/rust-{mode}`), but I think something like `builddir/{}/rust/scylla-{mode}` would be just as good. On the other hand, maybe a comment would suffice?

psarna replied:
a comment would suffice for me

wmitros replied:
Comment added in the rebase

psarna commented on rust/wasmtime_bindings/src/memory_creator.rs:
```diff
@@ -0,0 +1,97 @@
+            if let Ok(new_layout) = std::alloc::Layout::from_size_align(new_size, 64 * 1024) {
+                let new_ptr = match self.layout {
+                    Some(layout) => unsafe { std::alloc::realloc(self.ptr, layout, new_layout.size()) },
+                    None => unsafe { std::alloc::alloc(new_layout) },
```
just to confirm - is this whole thing done simply to force wasm to use host malloc/free for allocations, so that it gets automatically handled by the Seastar allocator?

wmitros replied:
Yes, otherwise the memory will be allocated with `mmap` that we don't override in seastar

psarna replied:
In this case, let's explicitly explain it in the comment, along with mentioning mmap. Technically, nothing guarantees that `std::alloc` doesn't switch to mmap one day

psarna replied:
(or, just use an unsafe malloc/free pair directly to be sure)

psarna replied:
Oh wait, it's an unsafe block anyway. In this case I strongly suggest you just forward-declare malloc, realloc, free, whatever's needed, and use them directly. Then we're 100% sure there won't be any mmap sneaking in.

And in case we have strict layout requirements, like a 64k page for wasm, perhaps instead of malloc we should just go with aligned_alloc (https://linux.die.net/man/3/aligned_alloc)

psarna replied:
Ref: https://github.com/rust-lang/rust/blob/c3605f8c8020dbbe8f0d1961c7b33c4c4b78ad0d/library/std/src/sys/unix/alloc.rs#L6-L24

with this implementation it uses aligned_alloc (or malloc for small allocations), which we properly override in Seastar indeed, and it will most likely stay that way due to the 64k alignment, because mmap doesn't exactly let you choose a multi-page alignment anyway. But in theory on on certain hugepage configurations it may be beneficial to use mmap over aligned_alloc, so in my opinion we should be prepared for that

wmitros replied:
Ok, I used  `aligned_alloc` and `free` in the new rebase. I've also added a comment about mmap.
I removed the `std::alloc::realloc` because it should probably not be used here anyway, considering it doesn't keep the alignment

psarna commented on lang/wasm.cc:
```diff
@@ -290,51 +195,60 @@ struct from_val_visitor {
+        if (eptr) {
+            co_await coroutine::return_exception_ptr(eptr);
+        }
+        co_await coroutine::maybe_yield();
```
:relieved:

avikivity replied:
I see you do yield here. Why are the others different? Need comments.

wmitros replied:
I don't yield in other places mostly because they should finish quickly, the other cases are also not used in future functions so it would require quite a bit more changes to yield there - I'll add comments.

psarna reviewed (commented):
Love it! @piodul FYI, a fair portion of Rust code to review

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/1918/

wmitros commented:
I've rebased onto master, added a comment and removed some unused code.

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/1981/

psarna commented:
@wmitros for starters, please also run `cargo clippy` (ref: https://github.com/rust-lang/rust-clippy#step-2-install-clippy) on the Rust source code, it will provide a few tips on how to make the code more idiomatic:
```
warning: explicit lifetimes given in parameter types where they could be elided (or replaced with `'_` if needed by type declaration)
  --> src/lib.rs:15:39
   |
15 |   #[cxx::bridge(namespace = "wasmtime")]
   |  _______________________________________^
16 | | mod ffi {
17 | |     enum ValKind {
18 | |         I32,
...  |
77 | |         type Fut<'a>;
78 | |         unsafe fn resume<'a>(self: &mut Fut<'a>) -> Result<bool>;
   | |________________________________________________________________^
   |
   = note: `#[warn(clippy::needless_lifetimes)]` on by default
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#needless_lifetimes

warning: explicit lifetimes given in parameter types where they could be elided (or replaced with `'_` if needed by type declaration)
  --> src/lib.rs:78:9
   |
78 |         unsafe fn resume<'a>(self: &mut Fut<'a>) -> Result<bool>;
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#needless_lifetimes

warning: explicit lifetimes given in parameter types where they could be elided (or replaced with `'_` if needed by type declaration)
  --> src/lib.rs:84:31
   |
84 |         ) -> Result<Box<Fut<'a>>>;
   |                               ^
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#needless_lifetimes

warning: this expression creates a reference which is immediately dereferenced by the compiler
   --> src/lib.rs:100:39
    |
100 |     let mut ctx = Context::from_waker(&waker);
    |                                       ^^^^^^ help: change this to: `waker`
    |
    = note: `#[warn(clippy::needless_borrow)]` on by default
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#needless_borrow

warning: the function `wasmtime::Store::new` doesn't need a mutable reference
   --> src/lib.rs:136:42
    |
136 |     let mut store = wasmtime::Store::new(&mut engine.wasmtime_engine, wasi);
    |                                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |
    = note: `#[warn(clippy::unnecessary_mut_passed)]` on by default
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#unnecessary_mut_passed

warning: `wasmtime_bindings` (lib) generated 5 warnings
    Finished dev [unoptimized + debuginfo] target(s) in 58.43s
```

avikivity commented on rust/Cargo.lock:
```diff
@@ -38,59 +392,1418 @@ dependencies = [
+ "errno-dragonfly",
+ "libc",
+ "winapi",
+]
```
Does rust have the equivalent of ccache that can store cached build results outside the build directory? So that `rm -rf build; ninja` will be fast?

I'm thinking of our builders which start from a clean directory each time.

piodul replied:
Mozilla's sccache works with rust: https://github.com/mozilla/sccache . Disclaimer: I have never used it personally so I can't tell how well it works in practice.

avikivity replied:
Ok. Let's start using it when we see rust compilation times having an impact.

wmitros replied:
The `incremental` option is supposed to help with rebuild times, but it does not help with the `rm -rf build; ninja` case and I only enabled it for the `dev` mode, as it might make the final binary less optimized.
I might just try this `sccache` because in the case of `rm -rf build; ninja` the rust compilation is very impactful (although this could have been expected, as rust is the only thing that is being rebuilt then)

wmitros replied:
I tried [sccache](https://github.com/mozilla/sccache) and it helps a lot with the `rm -rf build; ninja` case. I don't expect we want to add it as one of our dependencies, but explain how to use it in some docs.  I'll need to research it a bit more so maybe we can do this as a follow-up

avikivity replied:
We install ccache into the dbuild environment so we may as well install sscscscscache. Certainly it could be in a separate pull request.

avikivity replied:
I see it even supports C++, so if it's better in some way, we can replace ccache.

avikivity commented on rust/wasmtime_bindings/src/lib.rs:
```diff
@@ -0,0 +1,358 @@
+            }
+            Poll::<Result<wasmtime::Instance>>::Ready(Err(e)) => {
+                return Err(anyhow!("Failed to instantiate module: {:?}", e));
+            }
```
Why does instantiate_async() return a future?

If it's in order to yield, then aren't we wasting the opportunity by looping? We should return to the caller and let them yield too.

avikivity replied:
(instantiating an instance is less important than execution, since it's rarer, but still)

wmitros replied:
According to the documentation, we return a future here only to potentially yield when we "asynchronously invoke the wasm start function in case it calls any imported function which is an asynchronous host function". The only imports come from WASI, but I don't think they are used in the start function (actually I couldn't find the start function implementation - it is also possible that there is none).
If we also go with https://github.com/scylladb/scylladb/pull/11351#discussion_r962925721 no imports will be async at all, so it won't be possible that this future yields in the middle

avikivity commented on rust/wasmtime_bindings/src/lib.rs:
```diff
@@ -0,0 +1,358 @@
+        self.wasmtime_memory.size(&store.wasmtime_store)
+    }
+    fn grow(self: &mut Memory, store: &mut Store, delta: u64) -> Result<u64> {
+        self.wasmtime_memory.grow(&mut store.wasmtime_store, delta)
```
What does wasmtime use to allocate memory?

avikivity replied:
(Need to understand if it will be accounted for in seastar memory management)

psarna replied:
I skimmed through wasmtime code and it indeed mmaps a lot. Some of it is actual file-backed mapping, but often it's used as glorified malloc with MAP_ANONYMOUS. 

We could technically add an opt-in Seastar feature which makes extrernal mmap-based libraries more Seastar friendly:
1. Override `mmap` to fall back to original behavior, except when `MAP_ANONYMOUS` is set, in which case we just call aligned_alloc with page size boundary, which effectively gives us pages. Then, we can apply requested protection with a separate `mprotect` call.
2. `munmap` gets a little complicated, because I don't know any reliable way of checking if a page is file-backed or anonymous, except for examining `/proc/self/maps` output. But if we assume that the number of anonymous mappings is relatively small, perhaps adding a hashmap overhead is acceptable if it means that we regain the ability to allocate everything with a Seastar allocated. Then, we could just track our anonymous pages ourselves and override `munmap` behavior to a regular `free` if the page is tracked as "ours".

@avikivity Does it sound too obscure to you, or is it feasible? The main question is how many more mmap-ed libraries we're likely to integrate.

denesb replied:
Won't file-backed `mmap()` cause blocking memory accesses and stalls?

avikivity replied:
Not just stalls, also OOM since we already allocated all memory for our own use.

avikivity replied:
Overriding ::mmap is likely to end up with a lot of tears, it's too fundamental and too not designed to be overridden.

Can we patch wasmtime (upstream) not to use mmap? It seems like premature optimization. The system memory allocator will call mmap when the conditionals call for it.

Perhaps the issue is the need to make the memory executable later with mprotect() (typical jit operation is to first mmap with PROT_WRITE, then mprotect to PROT_EXEC.

If so, perhaps we can pass callbacks to wasmtime for mmap/mprotect/munmap so we can intercept them. The default callbacks will just call the system mmap/mprotect/munmap.

wmitros replied:
Completely removing `mmap` from wasmtime seems difficult, but maybe it is not essential after all. 
I inspected the wasmtime code and I found that the `mmap`s are used in:
- the default `Linear Memory` allocator - which we have replaced
- the `Pooling Memory Allocator` (where we preallocate all the memory for all the instances that we can later use) - which we do not use
- in the `copy-on-write` implementation of the initialization of `Linear Memories` - which is only used in the before-mentioned `Pooling Memory Allocator`
- for allocation of the stacks for asynchronous calls - this could probably be replaced with `malloc`
- for allocation of the alternate signal stack - this could also probably be replaced with `malloc`
- in compilation of a module from a file - which we currently do not use
- for allocation of for the jit code to be later `mprotect`ed with `PROT_EXEC` - the `mmap`s are all with `MAP_ANONYMOUS` but I'm not sure we can help it (according to the `mprotect` manual `mprotect()` is unspecified if it is applied to a of memory that was not obtained via `mmap()`)

denesb replied:
> for allocation of for the jit code to be later mprotected with PROT_EXEC - the mmaps are all with MAP_ANONYMOUS but I'm not sure we can help it (according to the mprotect manual mprotect() is unspecified if it is applied to a of memory that was not obtained via mmap())

All memory in the seastar allocator is obtained via `mmap()`. So we should be able to mprotect memory coming from it. But we have to be very careful to make sure the result of `mprotect` doesn't spill over to memory used by other objects and to remove it after not needed.

avikivity commented on rust/wasmtime_bindings/src/memory_creator.rs:
```diff
@@ -0,0 +1,95 @@
+        if new_size == 0 {
+            new_ptr = ptr::null_mut()
+        } else {
+            new_ptr = unsafe { aligned_alloc(WASM_PAGE_SIZE, new_size) };
```
This answers my previous question.

aligned_alloc isn't defined if new_size isn't a multiple of WASM_PAGE_SIZE, is it guaranteed by the caller?

wmitros replied:
When I look at our `aligned_alloc` overwrite it looks like we align the size up when necessary, but I see that in the linux specification it's as you described.
I'd expect the size here to be aligned, but I don't see it specified directly, so I think I'll just use memalign here

psarna replied:
(you could just align the new size yourself)

wmitros replied:
I'll align the size myself then

avikivity commented on lang/wasm.cc:
```diff
@@ -6,8 +6,6 @@
  */
 
-#ifdef SCYLLA_ENABLE_WASMTIME
-
```
Please check that the series continues building without wasmtime.

wmitros replied:
Wasmtime is now downloaded as part of the compilation of the .rs files, so it should never be missing, but it may not work correctly in ARM (it didn't work when we used libwasmtime.a) - I'll try to check this soon. If it doesn't work with ARM I'll bring the `SCYLLA_ENABLE_WASMTIME` back

avikivity replied:
Wait what? Didn't we add it to the toolchain?

In any case it may not be available on all architectures.

wmitros replied:
We did add Rust to the toolchain, but we didn't add all the crates that will be used in this patch.
Actually, we should probably discuss how do we want to deal with downloading the new crates.
Currently, the dependency will be downloaded if it's missing during the compilation (it won't be re-downloaded if it's already present).
If we don't want that, we could update the toolchain every time a dependency is changed.
We can also update the toolchain only periodically, and some dependencies could ever need to be downloaded during compilation.

I didn't really see benefits to the latter options because after the first build, there basically is no difference between them

wmitros replied:
Actually, where do we want to keep the downloaded crates? Cargo uses `$HOME/.cargo` by default.
To test if everything works with ARM [this](https://github.com/scylladb/scylla-pkg/issues/3062) would help a lot

avikivity commented on lang/wasm.cc:
```diff
@@ -19,190 +17,97 @@
+                    rets->push_i32(0);
+
+                    auto fut = wasmtime::get_func_future(store, *malloc_func, *argv, *rets);
+                    while(!fut->resume());
```
Code style: space after while, it isn't a function.

Also a comment indicating that we need to yield here.

If we set need_thread, then it may be enough to call seastar::thread::maybe_yield() here.

wmitros replied:
I can modify the `operator()` in the `init(_nullable)_arg_visitor` so that it returns a future, then we can use the `maybe_yield` like when executing the UDF

avikivity replied:
If this is just pushing arguments, maybe unneeded. But comments are definitely needed.

avikivity commented on lang/wasm.cc:
```diff
@@ -261,22 +166,22 @@ struct from_val_visitor {
+            argv->push_i32((int32_t)val.i64());
+            auto rets = wasmtime::get_val_vec();
+            auto free_fut = wasmtime::get_func_future(store, *free_func, *argv, *rets);
+            while(!free_fut->resume());
```
Ditto.

avikivity commented:
Looks good. Please add tests that prove we yield on a long loop.

piodul commented on rust/wasmtime_bindings/src/lib.rs:
```diff
@@ -0,0 +1,358 @@
+
+    loop {
+        match inst_fut.as_mut().poll(&mut ctx) {
+            Poll::<Result<wasmtime::Instance>>::Pending => {}
```
Is the turbofish needed here? It makes the line quite long and verbose. Would `Poll::Pending` work here instead?

piodul commented on rust/wasmtime_bindings/Cargo.toml:
```diff
@@ -0,0 +1,12 @@
+[dependencies]
+cxx = { version = "1.0.73", features = ["c++20"] }
+wasmtime = { version = "0.40.0", features = ["async"] }
+wasmtime-wasi = { version = "0.40.0", features = ["tokio"] }
```
Why do we need the `tokio` feature? We aren't running a tokio runtime, and we would rather avoid code which assumes running in the tokio runtime context (it will panic if it tries to access stuff from tokio).

wmitros replied:
The `tokio` feature was the only alternative to the `sync` feature, so it looked like using it was the only option for the async wasi, but now that I think about it, we might not need it here and just just the `sync` WASI - I'll test if the yielding works without it

piodul commented on rust/wasmtime_bindings/src/lib.rs:
```diff
@@ -0,0 +1,358 @@
+}
+
+fn get_abi(instance: &Instance, store: &mut Store, memory: &Memory) -> Result<u32> {
+    match instance
```
Style suggestion: the function returns a `Result<>`, so you can use the `?` operator and some methods on `Option` to make the code flatter. For example:

```rust
let export = instance
    .wasmtime_instance
    .get_export(&mut store.wasmtime_store, "_scylla_abi")
    .ok_or_else(|| anyhow!("ABI version export not found - please export `_scylla_abi` in the wasm module"))?;

let global = global
    .get(&mut store.wasmtime_store)
    .ok_or_else(|| anyhow!("Exported object _scylla_abi is not a global"))?;

// ...
```

Other functions would benefit from this conversion, too.

wmitros replied:
I also added some `map_err` in the same style in the rebase

piodul commented on replica/database.cc:
```diff
@@ -361,7 +361,7 @@ database::database(const db::config& cfg, database_config dbcfg, service::migrat
     , _shared_token_metadata(stm)
     , _sst_dir_semaphore(sst_dir_sem)
-    , _wasm_engine(std::make_unique<wasm::engine>())
+    , _wasm_engine(std::make_unique<rust::Box<wasmtime::Engine>>(wasmtime::create_engine()))
```
Isn't `rust::Box` something like a `std::unique_ptr` already? Would it be possible to just use `rust::Box` directly and remove the double indirection?

wmitros replied:
I tried using the `rust::Box` alone in the rebase and looks like it works for now

piodul commented on lang/wasm.cc:
```diff
@@ -290,51 +195,60 @@ struct from_val_visitor {
             "f32",
             "f64",
             "v128",
-            "externref",
             "funcref",
+            "externref",
```
Why was the order changed here?

wmitros replied:
It matched the order of enum elements in `lang/wasmtime.hh` https://github.com/scylladb/scylladb/blob/41b91e36324fbd373d3fe6ca1006342f663e2568/lang/wasmtime.hh#L492-L497 and now it matches the order in the wasmtime crate https://docs.rs/wasmtime/latest/wasmtime/enum.Val.html

psarna commented on rust/wasmtime_bindings/src/lib.rs:
```diff
@@ -0,0 +1,358 @@
+    wasmtime_engine: wasmtime::Engine,
+}
+
+fn create_engine() -> Result<Box<Engine>> {
```
am I right to assume that Result<T> gets translated by cxx to produce a function that returns a T if the result is Ok, and throws an exception in case of Err?

piodul replied:
Yup: https://cxx.rs/binding/result.html#returning-result-from-rust-to-c

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/2104/

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/2124/

wmitros commented:
In the last rebase(s) I added the test for yielding.

For now I went with the "download if missing" approach for the problem from https://github.com/scylladb/scylladb/pull/11351#discussion_r962898100 but it causes CI to fail now because I changed the location where rust crates are downloaded from `$builddir/{mode}/.cargo` to the default location which is `$HOME/.cargo`.

Neither of these locations is good, because if we choose the old location the crates will have to be downloaded each time the build directory is deleted, and if we choose the new location the CI will fail.

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/2127/

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/2259/

wmitros commented:
I added a new commit that introduces some wasm UDF configuration and fixed some conflicts in the rebase

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/2266/

psarna commented:
side note: I think that the title of this PR should be modified to mention UDF explicitly, "Add custom rust wasmtime bindings" is a little too generic and doesn't express that we actually rewritten the whole UDF layer to Rust, which is a big thing!

wmitros commented:
In the last version I fixed some conflicts and renamed the PR to be more specific

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/2273/

wmitros commented:
The test for yielding during a long UDF execution is slower on jenkins, but we assert that the time before yielding is <1ms, so I modified the wasmtime fuel that's consumed before yielding to an even lower amount

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/2279/

wmitros commented:
The test should pass now even on the slower machines

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/2284/

wmitros commented:
The test should now pass on even slower machines

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/2300/

wmitros commented:
I've adjusted the time limit for the slow machines, and added some more configuration options

wmitros commented:
The test should now pass even on mechanical computers

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/2325/

denesb commented:
> During the development of this patch I realized an issue that isn't linked to this patch, but to Rust as a whole - the behavior on OOM errors. Stable Rust `aborts()` on such an error and does not provide methods that allow writing code with a fallible allocations - there are `try_reserve` methods for some Collections but this method is missing for e.g. the `Box` type, which is crucial for us. This could at least partially be helped by `nightly` Rust - the `Box::try_new` is available there, but I would still need to check if the remaining code can be made safe then

Does this mean that user-provided code *has* to use the subset of methods that explicitly allow for failure? What happens if the user doesn't know/care or simply forgets and uses the methods that trigger abort on bad-alloc? Will this bring the scylladb node down? Or maybe I'm misunderstanding this completely and it is only us who have to be careful when writing the bindings?

wmitros commented:
> Does this mean that user-provided code _has_ to use the subset of methods that explicitly allow for failure? What happens if the user doesn't know/care or simply forgets and uses the methods that trigger abort on bad-alloc? Will this bring the scylladb node down? Or maybe I'm misunderstanding this completely and it is only us who have to be careful when writing the bindings?

The user's code is compiled to WASM and this issue does not concern WASM (the only real allocation in WASM is growing the associated WASM Linear Memory, and if this fails, the `grow` function simply returns -1) so there are no problems for the users.

The actual issue is that it's not enough to be careful to avoid OOM aborts, the Rust language simply doesn't make it possible to do so. In particular, even if we were able to handle OOM in our bindings, the program could still abort in one of the Rust crates that we were using.

Luckily, the main source of large allocations will be the allocation of WASM Linear Memories which does not cause aborts in our bindings. For most of its bigger allocations, wasmtime also often uses `mmap` which does allow error handling, though there are separate [issues](https://github.com/scylladb/scylladb/pull/11351#discussion_r965657980) with that

denesb commented:
I'm not sure I follow. You say that rust code cannot handle OOM but that is ok as the only big allocations are in WASM which does handle this. I'm confused, won't rust be compiled to WASM so WASM failing to aquire more memory will lead to some rust allocation failing, no?
Also, how does such an abort in rust look like? Since we control compilation of rust and running the resulting code, can we intercept such abort attempts and just convert it to a failed UDF/UDA/whatever run?

Note that using `mmap()` itself might be problematic. Seastar already `mmaps()` $all_physycal_memory - $reserve` memory and AFAIK also "faults them in" (forces kernel to assign backing memory). Additional `mmap()` calls will eat into the reserve seastar left for the operating system, which is dangerous -- we will end up violating our promise of leaving `$reserve` memory to the OS and other apps (scylla manager, nodetool, jmx, etc.).

wmitros commented:
> I'm not sure I follow. You say that rust code cannot handle OOM but that is ok as the only big allocations are in WASM which does handle this. I'm confused, won't rust be compiled to WASM so WASM failing to aquire more memory will lead to some rust allocation failing, no?

When compiling Rust to WASM, the Rust aborts are actually compiled to WASM traps, which we later get (after some translation done by the WASM runtime and the cxx-rust bridge) as exceptions in the Scylla C++ code. We can deal with exceptions, so even if user's code aborted, that does not cause Scylla to crash.

An important thing that I did not describe clearly is that some allocations can be performed without aborting in Rust. When I said that Rust does not allow writing code not subject to aborting on allocation failure I meant that there are some data types that have this problem, not that all allocations do. For example, we can try to create a Rust vector using `Vec::try_reserve()` but we can't do that with a shared pointer `Box`.
In particular, for some allocations we can provide our own allocator - this is the case with WASM Linear Memories. 
When growing a Memory, we actually use our own allocator from Seastar. If it fails (i.e. returns a null pointer) the result of the WASM `memory.grow` instruction is -1 and the program continues. 

If the WASM bytecode was compiled from Rust, the `memory.grow` will probably be followed by a comparison of the returned value to -1 and an optional trap, which will appear as an exception in our C++ code.

> Also, how does such an abort in rust look like? Since we control compilation of rust and running the resulting code, can we intercept such abort attempts and just convert it to a failed UDF/UDA/whatever run?

The abort in Rust is basically the same as in C++ - it's the SIGABRT signal. Maybe we could register a signal handler and set up some recovery path using setjmp/longjmp but I'm not sure that's the best solution.

> Note that using `mmap()` itself might be problematic. Seastar already `mmap()`s `$all_physical_memory - $reserve` memory and AFAIK also "faults them in" (forces kernel to assign backing memory). Additional `mmap()` calls will eat into the reserve seastar left for the operating system, which is dangerous -- we will end up violating our promise of leaving `$reserve` memory to the OS and other apps (scylla manager, nodetool, jmx, etc.).

I agree that it's problematic and we should probably discuss this further. @psarna had same ideas about improving this, and I see you've already responded so let's continue this topic there.

raphaelsc commented on db/config.cc:
```diff
@@ -907,6 +907,12 @@ db::config::config(std::shared_ptr<db::extensions> exts)
     , task_ttl_seconds(this, "task_ttl_in_seconds", liveness::LiveUpdate, value_status::Used, 10, "Time for which information about finished task stays in memory.")
     , cache_index_pages(this, "cache_index_pages", liveness::LiveUpdate, value_status::Used, true,
         "Keep SSTable index pages in the global cache after a SSTable read. Expected to improve performance for workloads with big partitions, but may degrade performance for workloads with small partitions.")
+    , wasm_cache_size(this, "wasm_cache_size", value_status::Used, 128 * 1024 * 1024, "Maximum total size of all WASM instances stored in the cache")
```
is it per shard? shouldn't we make it a percentage of total memory available? then it automatically adjust based on memory available for each shard.

avikivity commented:
> During the development of this patch I realized an issue that isn't linked to this patch, but to Rust as a whole - the behavior on OOM errors. Stable Rust `aborts()` on such an error and does not provide methods that allow writing code with a fallible allocations - there are `try_reserve` methods for some Collections but this method is missing for e.g. the `Box` type, which is crucial for us. This could at least partially be helped by `nightly` Rust - the `Box::try_new` is available there, but I would still need to check if the remaining code can be made safe then

Is this for rust running in a udf, or rust linked to the scylla binary?

avikivity commented on rust/wasmtime_bindings/src/lib.rs:
```diff
@@ -0,0 +1,347 @@
+    // The following configuration settings make wasmtime allocate only as much memory as it needs
+    config.static_memory_maximum_size(0);
+    config.dynamic_memory_reserved_for_growth(0);
+    config.dynamic_memory_guard_size(0);
```
Without the guard, will it emit bound checks?

wmitros replied:
Yes. If we didn't provide our own memory allocator, the memory would be allocated using mmap by default and the guard would actually allow eliding some of the checks - the guard region would remain unmapped and would trigger a segfault if accessed.

avikivity commented on rust/wasmtime_bindings/src/memory_creator.rs:
```diff
@@ -0,0 +1,96 @@
+use anyhow::{anyhow, Result};
+use std::{cmp, ptr, slice, u32};
```
Copyright/license blurbs needed in all new files.

wmitros replied:
Ok

wmitros commented:
> > During the development of this patch I realized an issue that isn't linked to this patch, but to Rust as a whole - the behavior on OOM errors. ...
> 
> Is this for rust running in a udf, or rust linked to the scylla binary?

It's the rust linked to the scylla binary. I tried to explain this a bit better in my answers for @denesb above

piodul commented:
@wmitros @denesb about the issue with abort on OOM: there seems to be a plan to add an option so that in case of OOM a panic will be thrown: https://github.com/rust-lang/rust/issues/43596 . It should be possible to catch such a panic with `std::panic::catch_unwind`. I'm not sure how panics are translated across the Rust-C++ boundary by `cxx`, though.

At the same time, there are some reasonable concerns about whether existing libraries (including wasmtime) handle panicking on OOM correctly. Until recently, it was reasonable to assume that OOM=abort, always. Turning this option on could be especially dangerous with libraries that use unsafe code (like wasmtime). As bad as it sounds, leaving the old behavior might be a better idea, unless the library is written with OOM=panic in mind and tested against it.

It looks like an option to enable panic-on-oom is already available in nightly (`-Z oom=panic` seems to work).

denesb commented:
Wichever we choose, we have to make sure we have a good automated test for the OOM scenario to ensure it is handled reasonably (scylla doesn't crash).

wmitros commented:
The patch now needs https://github.com/scylladb/seastar/pull/1328 to be merged to successfully build.

In the last rebase I added a test confirming that even if an allocation of a WASM Memory fails the program will not abort 
Additionally, a system for limiting the total memory used by the compiled WASM Modules has been implemented - as we previously discussed, we need to do this separately because the executable code for the modules is allocated using mmap and not our seastar allocator.

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/3436/

wmitros commented:
I have verified with `strace` that we're tracking the memory accordingly with the `mmap` and `munmap` calls that are happening. The wasm runtime allocates the memory in 2 cases: for executable code and for separate stacks for asynchronous UDFs. We can now track them both.
I also verified that the `mmap` is not used for WASM Memories - we allocate them using seastar `malloc`, but, as a result, we cannot elide any bounds checks when accessing the memory.

Both the change for using `mmap` and the change for testing `oom` turned out quite complicated so please look at the comments that have been added.

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/3464/

wmitros commented:
This has probably gotten buried in your notifications so let me @ you @avikivity

avikivity commented:
> I have verified with `strace` that we're tracking the memory accordingly with the `mmap` and `munmap` calls that are happening. The wasm runtime allocates the memory in 2 cases: for executable code and for separate stacks for asynchronous UDFs. We can now track them both. I also verified that the `mmap` is not used for WASM Memories - we allocate them using seastar `malloc`, but, as a result, we cannot elide any bounds checks when accessing the memory.

That's fine. The bounds check is not trivial but bearable, and much allows for much better resource control.

I guess the separate stack is per-instance? Can it be moved to malloc too?


> 
> Both the change for using `mmap` and the change for testing `oom` turned out quite complicated so please look at the comments that have been added.

avikivity commented on lang/wasm.cc:
```diff
@@ -290,51 +197,60 @@ struct from_val_visitor {
+    try {
+        ctx.module = wasmtime::create_module(ctx.engine_ptr, rust::Str(script.data(), script.size()));
+        auto store = wasmtime::create_store(ctx.engine_ptr);
+        auto inst = create_instance(ctx.engine_ptr, **ctx.module, *store);
```
Shouldn't creation of an module and store/instance be separated? We create many instances per module.

wmitros replied:
The module is indeed created only once. In this case we create an instance (and a Func) alongside it to confirm that the UDF is usable (the compilation of the module can succeed for basically any valid WASM code, which may or may not contain our function). 
After confirming this, the instance is destroyed and other instances are created only when we need them for UDF executions

avikivity replied:
Ok, please add a comment to explain.

wmitros replied:
Comment added

avikivity commented on test/boost/wasm_test.cc:
```diff
@@ -0,0 +1,84 @@
+    while (!stop) {
+        auto start = system_clock::now();
+        stop = fut->resume();
+        max_duration = std::max(max_duration, system_clock::now() - start);
```
This is likely to not work when running in parallel with other tests on overloaded machines.

Perhaps you can make it more robust by capturing CPU time (clock_gettime(CLOCK_THREAD_CPUTIME_ID)) and discarding measurements where the cpu time and real time differ by too much.

You can try to reproduce the problem by running many tight loops on the same cpu as the test, just to make sure you indeed fix the problem.

wmitros replied:
Sounds good to me. I did manage to fail it using the tight loops approach and using the CLOCK_THREAD_CPUTIME_ID seems to help.

avikivity replied:
> > I guess the separate stack is per-instance? Can it be moved to malloc too?
> 
> Yes, it's per instance and it's only using space when the instance is actually used (not while it's in the cache). There is currently no option to use stacks allocated using malloc, but it sounds like it would be a smaller problem to contribute to wasmtime with this change, than with the change of allocations of jit code

Eventually we'll want to contribute fragmented memory, where memory is an array of equally-sized blocks. It means loads/stores will translate to several machine instructions, but it does make grow O(1) (so it's useful not just to us with our fragmentation phobia, but to anyone who cares about latency).

avikivity commented on test/boost/wasm_alloc_test.cc:
```diff
@@ -0,0 +1,94 @@
+#include "seastar/core/reactor.hh"
+#include <seastar/testing/test_case.hh>
+#include <seastar/core/coroutine.hh>
+
```
Missing copyright/license. Please check other new files too.

Better to consolidate wasm tests in fewer files, to reduce bloat on disk as our executables are large.

wmitros replied:
I added the license to the files where it was missing. I did try to put all wasm-related tests in one file, but for the reasons described in this test file, it was not possible when testing traps in wasm. If we make more tests for wasm traps, we may need to prepare a workaround for this to avoid creating a new test file for each of such tests

avikivity commented:
Looks good, but I only gave it a cursory review, I'm not used to reviewing Rust. @piodul your help would be appreciated.

Suggestion for another test: a wasm function that purposefully oversteps its memory, to see that we trap. Can be a follow-up.

I'm disappointed that traps are communicated as signals (they are slow), but it's not too terrible.

wmitros commented:
> I guess the separate stack is per-instance? Can it be moved to malloc too?

Yes, it's per instance and it's only using space when the instance is actually used (not while it's in the cache). 
There is currently no option to use stacks allocated using malloc, but it sounds like it would be a smaller problem to contribute to wasmtime with this change, than with the change of allocations of jit code

wmitros commented:
The last changes are included in the last force-push, the other one is a rebase on master

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/3577/

piodul commented on rust/wasmtime_bindings/src/memory_creator.rs:
```diff
@@ -0,0 +1,104 @@
+    fn maximum_byte_size(&self) -> Option<usize> {
+        self.maximum_size
+    }
+    fn grow_to(&mut self, new_size: usize) -> Result<()> {
+        let new_size_aligned = (new_size + WASM_PAGE_SIZE - 1) & !(WASM_PAGE_SIZE - 1);
```
What is the maximum value of the `new_size` parameter? If it can be close to `usize::MAX` then the expression for `new_size_aligned` can overflow to 0. Perhaps we should calculate `(new_size + WASM_PAGE_SIZE - 1)` using the `checked_add` function (calculate `new_size.checked_add(WASM_PAGE_SIZE - 1)` and return Err if the result was None).

wmitros replied:
I don't see any external factors limiting the `new_size` parameter, but even though it makes little sense to use it with a value close to `usize::MAX`, I agree that it's better to be consistent and use the `checked_add` method for this.

piodul commented on rust/wasmtime_bindings/src/memory_creator.rs:
```diff
@@ -0,0 +1,104 @@
+        }
+        let copy_size = cmp::min(self.size, new_size_aligned);
+        unsafe {
+            slice::from_raw_parts_mut(new_ptr, copy_size)
+                .copy_from_slice(slice::from_raw_parts(self.ptr, copy_size))
```
The use of `from_raw_parts`/`from_raw_parts_mut` is undefined behavior if the pointer that you use to construct the slice from is null, see here: https://doc.rust-lang.org/std/slice/fn.from_raw_parts_mut.html . You should just skip the copy if any of the pointers is null or `copy_size == 0`.

piodul replied:
By the way, you could probably use [`std::ptr::copy_nonoverlapping`](https://doc.rust-lang.org/std/ptr/fn.copy_nonoverlapping.html) here which will result in shorter code. It has the same restrictions with regards to the pointers, though.

wmitros replied:
I added both the restrictions and the  [`std::ptr::copy_nonoverlapping`](https://doc.rust-lang.org/std/ptr/fn.copy_nonoverlapping.html)

wmitros commented on rust/wasmtime_bindings/src/test_memory_creator.rs:
```diff
@@ -0,0 +1,133 @@
+    }
+}
+
+unsafe impl LinearMemory for TestScyllaLinearMemory {
```
I managed to mostly fix this in the rebase.  The new behaviour of the `TestScyllaLinearMemory::grow_to` is now slightly different - the change isn't really important but I described it below anyway.
Because the duplication of the logic of deciding when we even need to perform any allocations has been removed, now, instead of inserting the failed alloc after we decided that it was needed using the mentioned logic, we now perform the alloc using the `ScyllaLinearMemory` and depending on the result we insert the failed alloc afterwards. 
A side effect of this is that the `TestScyllaLinearMemory::grow_to` will succeed even though it returns an `Err`, leaving the `TestScyllaLinearMemory` in an unexpected state, but we aren't planning to use it anyway later, so there's probably no harm in it.

piodul replied:
I'm not sure whether it's a problem either. The documentation of `LinearMemory` is not clear on what should happen with memory when grow_to fails ("Returns Err if memory can’t be grown by the specified amount of bytes").

If you are concerned about grow_to changing the memory pointer on failure, I guess you could split the current `ScyllaLinearMemory::grow_to` into two functions:

```rust
/// Validate and adjust the size to be allocated.
/// Returns None if realllocation is not needed
pub fn grow_prepare(&self, new_size: usize) -> Result<Option<usize>>;

/// Actually performs the reallocation.
/// Assumes that the size was returned from grow_prepare.
pub fn grow_perform(&mut self, new_size: usize) -> Result<()>;
```

Then you could use them to implement grow_to in both `ScyllaLinearMemory` and `TestScyllaLinearMemory`.

wmitros replied:
I think I'll leave it as it is now. As you mentioned, LinearMemory is not clear on what should happen with memory when grow_to fails, and there is no reason for us to introduce a new rule for it - it would be even worse if we started depending on it and be surprised later.

If I were changing this, I would probably use your method with `grow_prepare` and `grow_perform`, however.

wmitros commented on lang/wasm_instance_cache.cc:
```diff
@@ -102,7 +108,14 @@ seastar::future<instance_cache::value_type> instance_cache::get(const db::functi
+                auto ex = std::current_exception();
+                entry->instance = std::nullopt;
+                entry->mutex.unlock();
+                std::rethrow_exception(ex);
```
Ok

wmitros commented on rust/wasmtime_bindings/src/memory_creator.rs:
```diff
@@ -0,0 +1,104 @@
+        if new_size_aligned == self.size {
+            return Ok(());
+        }
+        let max_size = self.maximum_size.unwrap_or(u32::MAX as usize);
```
I agree, `u32::MAX + 1` is a lot better here

wmitros commented on rust/wasmtime_bindings/src/memory_creator.rs:
```diff
@@ -0,0 +1,104 @@
+            return Ok(());
+        }
+        let max_size = self.maximum_size.unwrap_or(u32::MAX as usize);
+        assert!(new_size_aligned <= max_size);
```
Ok

wmitros commented on docs/dev/rust.md:
```diff
@@ -7,32 +7,43 @@ shows how to use them in Scylla.
+2. Add `new_pkg = { path = "new_pkg", version = "0.1.0" }` to the dependencies list in `rust/Cargo.toml`
+3. Add `extern crate new_pkg;` to `rust/src/lib.rs`
+4. Configure your package in `new_pkg/Cargo.toml` and write your Rust code in `new_pkg/src/lib.rs` (and other `new_pkg/src/*.rs` files)
+5. To export a function `fn foo(x: i32) -> i32` in the namespace `xyz`, add its declaration to `new_pkg/src/lib.rs` as follows:
```
Ok

wmitros commented on docs/dev/rust.md:
```diff
@@ -7,32 +7,43 @@ shows how to use them in Scylla.
 1. Create a new package in the `rust` directory using `cargo new new_pkg --lib`
-2. Modify crate type to a static library by adding
+2. Add `new_pkg = { path = "new_pkg", version = "0.1.0" }` to the dependencies list in `rust/Cargo.toml`
+3. Add `extern crate new_pkg;` to `rust/src/lib.rs`
```
The `rust/src/lib.rs` file is used to combine all the rust packages into a single library and `extern crate` did help me with that. I see that I could have used `pub use new_pkg::*` instead, do you think that would be better?

piodul replied:
Right, but when you add a rust package you already have to specify it in the `Cargo.toml` of the `rust` crate (see point 2.) - so why is it also needed to add `extern crate new_pkg`?

Is it about linking? I.e. maybe cargo can decide not to link with `new_pkg` if it is not actually referred to in the `rust` crate?

wmitros replied:
> Is it about linking? I.e. maybe cargo can decide not to link with `new_pkg` if it is not actually referred to in the `rust` crate?

Looks like that is the case. At least in my testing, if I removed the `extern crate new_pkg` from `rust/src/lib.rs` I started getting `undefined symbol` errors

wmitros commented on docs/dev/rust.md:
````diff
@@ -7,32 +7,43 @@ shows how to use them in Scylla.
 ```
-6. Add `"rust/new_pkg/src/lib.rs"` to the `rusts` list in `configure.py`
-7. Include the `rust/new_pkg.hh` header and use `rust::foo()` in the selected c++ file.
+6. Add `new_pkg/src/lib.rs` to the `configure.py` dependencies where you'll need the Rust exports
+7. Include the `rust/new_pkg.hh` header and use `xyz::foo()` in the selected c++ file.
````
I added an option to put the cxx::bridge into other files to the `configure.py`.
I mentioned this option as an annotation and left the base guide unchanged because you need to satisfy some extra conditions to be able to use it and it would make the guide too complicated if we added exact descriptions of all of them to it.

wmitros commented on configure.py:
```diff
@@ -1943,14 +1943,17 @@ def configure_abseil(build_dir, mode, mode_config):
-                    objs.append('$builddir/' + mode + '/rust/release/' + lib)
+                    has_rust = True
+                    obj = dep.replace('/src/lib.rs', '.o').replace('rust/','')
+                    objs.append('$builddir/' + mode + '/gen/rust/' + obj)
```
The `.o` file that we are linking here is a result of compiling the `.cc` file generated from the cxx::bridge. For example, if we have 
```
/// lib.rs
#[cxx::bridge(namespace = "rust")]
mod ffi {
    extern "Rust" {
        fn inc(x: i32) -> i32;
    }
}
```
in a Rust `inc` package, we generate
```
/// inc.hh
namespace rust {
::std::int32_t inc(::std::int32_t x) noexcept;
} // namespace rust
```
and
```
/// inc.cc
namespace rust {
extern "C" {
::std::int32_t rust$cxxbridge1$inc(::std::int32_t x) noexcept;
} // extern "C"

::std::int32_t inc(::std::int32_t x) noexcept {
  return rust$cxxbridge1$inc(x);
}
} // namespace rust
```
We link to the result of compiling crates additionally to this, but that gives us the `librust_combined.a` file which actually contains only the `rust$cxxbridge1$inc` functions.

piodul replied:
OK, got it, thanks.

wmitros commented on rust/wasmtime_bindings/src/lib.rs:
```diff
@@ -0,0 +1,345 @@
+}
+
+impl Memory {
+    fn data(self: &Memory, store: &Store) -> *mut u8 {
```
I think I copied the definitions from the `mod ffi` block. I fixed this in the rebase

wmitros commented on rust/wasmtime_bindings/src/lib.rs:
```diff
@@ -0,0 +1,345 @@
+    let mut linker = wasmtime::Linker::new(&engine.wasmtime_engine);
+    if let Err(e) = wasmtime_wasi::add_to_linker(&mut linker, |s| s) {
+        return Err(anyhow!("Failed to add wasi to linker: {:?}", e));
+    }
```
Ok

wmitros commented on test/boost/wasm_test.cc:
```diff
@@ -0,0 +1,102 @@
+        clock_gettime(CLOCK_THREAD_CPUTIME_ID, &cpu_end);
+        auto cpu_duration = duration_cast<nanoseconds>(seconds(cpu_end.tv_sec) + nanoseconds(cpu_end.tv_nsec) - seconds(cpu_start.tv_sec) - nanoseconds(cpu_start.tv_nsec));
+        auto curr_duration = system_clock::now() - start;
+        if (duration_cast<microseconds>(curr_duration).count() < duration_cast<microseconds>(cpu_duration).count() * 10 ) {
```
Ok

wmitros commented on rust/wasmtime_bindings/src/memory_creator.rs:
```diff
@@ -91,10 +99,11 @@ unsafe impl wasmtime::MemoryCreator for ScyllaMemoryCreator {
         assert_eq!(guard_size_in_bytes, 0);
         assert!(reserved_size_in_bytes.is_none());
         assert!(!ty.is_64());
+        let max_size = std::cmp::min(self.max_scylla_size, maximum.unwrap_or(u32::MAX as usize));
```
Again, the `u32::MAX + 1` seems more appropriate

wmitros commented on lang/wasm.cc:
```diff
@@ -286,6 +288,7 @@ seastar::future<bytes_opt> run_script(const db::functions::function_name& name,
         func_inst = ctx.cache->get(name, arg_types, ctx).get0();
         ret = wasm::run_script(ctx, *func_inst->instance->store, *func_inst->instance->instance, *func_inst->instance->func, arg_types, params, return_type, allow_null_input).get0();
     } catch (const wasm::instance_corrupting_exception& e) {
+        ctx.cache->remove_module_ref(func_inst->module);
```
I remember that I had a problem with handling this just on construction and destruction, but after trying again, it looks like it's possible and probably even looks better.

wmitros commented on lang/wasm_instance_cache.cc:
```diff
@@ -100,9 +119,23 @@ seastar::future<instance_cache::value_type> instance_cache::get(const db::functi
     return f.then([this, entry, &ctx] {
+        // When the instance leaves the cache, it should be ready to be used. For
+        // that, we need to make sure that there is enough free memory for the
+        // wasm runtime stack, that is allocated at the start of the UDF executaion,
```
Ok

wmitros commented on rust/wasmtime_bindings/src/lib.rs:
```diff
@@ -116,15 +127,53 @@ fn create_instance(engine: &Engine, module: &Module, store: &mut Store) -> Resul
+}
+
+impl Module {
+    fn raw_size(self: &Module) -> usize {
```
Ok

wmitros commented on rust/wasmtime_bindings/src/lib.rs:
```diff
@@ -94,8 +101,12 @@ fn create_instance(engine: &Engine, module: &Module, store: &mut Store) -> Resul
     let mut inst_fut =
-        Box::pin(linker.instantiate_async(&mut store.wasmtime_store, &module.wasmtime_module));
+        Box::pin(linker.instantiate_async(&mut store.wasmtime_store, module.wasmtime_module.as_ref().unwrap()));
+
```
Thanks, I'll try to keep that in mind. I included your suggestion and seems like it works

wmitros commented on test/boost/wasm_alloc_test.cc:
```diff
@@ -0,0 +1,102 @@
+            continue;
+        } catch (const rust::Error& e) {
+            errors_during_execution++;
+            continue;
```
Makes sense, I added it in the rebase.

wmitros commented on rust/wasmtime_bindings/Cargo.toml:
```diff
@@ -0,0 +1,15 @@
+wasmtime = { version = "1.0.0", default-features = false, features = [
+    "async",
+    "wat",
+    "cranelift"] }
```
It does look better after the change, thanks.

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/3673/

wmitros commented:
I included fixes for @piodul's review in the first push and rebased on top of master in the second

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/3674/

wmitros commented:
Actually, I didn't address one comment in the previous rebase, I added it in the last one

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/3675/

piodul commented on rust/wasmtime_bindings/src/memory_creator.rs:
```diff
@@ -0,0 +1,123 @@
+            }
+        }
+        let copy_size = cmp::min(self.size, new_size_aligned);
+        if copy_size > 0 && !new_ptr.is_null() && !self.ptr.is_null() {
```
Nit: `copy_size > 0` implies `!new_ptr.is_null() && !self.ptr.is_null()`, right? I don't think the null checks are required.

wmitros replied:
Looks like it, I replaced them with a comment so that no one has doubts about it

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/3694/

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/3703/

nyh commented:
There seems to be a build error:
```
2023-01-04T17:57:21.586Z] FAILED: build/debug/main.o 
[2023-01-04T17:57:21.586Z] clang++ -MD -MT build/debug/main.o -MF build/debug/main.o.d -I/jenkins/workspace/releng/Scylla-CI/scylla/seastar/include -I/jenkins/workspace/releng/Scylla-CI/scylla/build/debug/seastar/gen/include -U_FORTIFY_SOURCE -DSEASTAR_SSTRING -Werror=unused-result -fstack-clash-protection -fsanitize=address -fsanitize=undefined -fno-sanitize=vptr -DSEASTAR_API_LEVEL=6 -DSEASTAR_SCHEDULING_GROUPS_COUNT=16 -DSEASTAR_DEBUG -DSEASTAR_DEFAULT_ALLOCATOR -DSEASTAR_SHUFFLE_TASK_QUEUE -DSEASTAR_DEBUG_SHARED_PTR -DSEASTAR_TYPE_ERASE_MORE -DFMT_SHARED -DSEASTAR_BROKEN_SOURCE_LOCATION -I/usr/include/p11-kit-1  -DDEBUG -DSANITIZE -DDEBUG_LSA_SANITIZER -DSCYLLA_ENABLE_ERROR_INJECTION -Og -DSCYLLA_BUILD_MODE=debug -g -gz -iquote. -iquote build/debug/gen --std=gnu++20  -ffile-prefix-map=/jenkins/workspace/releng/Scylla-CI/scylla=.  -march=westmere -DBOOST_TEST_DYN_LINK   -DNOMINMAX -DNOMINMAX -fvisibility=hidden  -Wall -Werror -Wno-mismatched-tags -Wno-tautological-compare -Wno-parentheses-equality -Wno-c++11-narrowing -Wno-sometimes-uninitialized -Wno-return-stack-address -Wno-missing-braces -Wno-unused-lambda-capture -Wno-overflow -Wno-noexcept-type -Wno-error=cpp -Wno-ignored-attributes -Wno-overloaded-virtual -Wno-unused-command-line-argument -Wno-defaulted-function-deleted -Wno-redeclared-class-member -Wno-unsupported-friend -Wno-unused-variable -Wno-delete-non-abstract-non-virtual-dtor -Wno-braced-scalar-init -Wno-implicit-int-float-conversion -Wno-delete-abstract-non-virtual-dtor -Wno-uninitialized-const-reference -Wno-psabi -Wno-narrowing -Wno-array-bounds -Wno-nonnull -Wno-uninitialized -Wno-error=deprecated-declarations -DXXH_PRIVATE_API -DSEASTAR_TESTING_MAIN -DFMT_DEPRECATED_OSTREAM -DHAVE_LZ4_COMPRESS_DEFAULT  -c -o build/debug/main.o main.cc
[2023-01-04T17:57:21.587Z] main.cc:463:13: error: no member named 'reserve_additional_memory' in 'seastar::app_template::config'
[2023-01-04T17:57:21.587Z]     app_cfg.reserve_additional_memory = 50 * 1024 * 1024;
[2023-01-04T17:57:21.587Z]     ~~~~~~~ ^
[2023-01-04T17:57:21.587Z] 1 error generated.
```

Weird, is this something you even changed in your patch?

wmitros commented:
> Weird, is this something you even changed in your patch?

Yes, this is something I added to seastar and it became usable only recently, with the last seastar update.
Looks like I didn't use the latest master branch for the CI. It should build when I rebase

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/3707/

wmitros commented:
> > CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/3707/
> 
> This time it seems to be a legitimate failure in the `test_fib_called_on_null`.
> 
> ```
> thread '<unnamed>' panicked at 'attempt to subtract with overflow', wasmtime_bindings/src/lib.rs:181:9
> ...
>   12:         0x1262fba3 - wasmtime$cxxbridge1$Module$remove_user
>                                at /jenkins/workspace/releng/Scylla-CI/scylla/rust/wasmtime_bindings/src/lib.rs:47:42
>   13:         0x11892e86 - _ZN4wasm14instance_cache17remove_module_refERN8wasmtime6ModuleE
>                                at ./lang/wasm_instance_cache.cc:240:12
> ...
> ```
> 
> Looks like the reference count for a module went below zero. In debug mode, Rust panics on under/overflow.

Yes, I found that while adding the "handle" for modules there was one place where I didn't remove the manual reference counting. I removed it and now the references are counted only in the handle constructor/destructor.

piodul commented on lang/wasm_instance_cache.hh:
```diff
@@ -113,6 +130,39 @@ public:
 
     void remove(const db::functions::function_name& name, const std::vector<data_type>& arg_types) noexcept;
 
+    // Wasmtime instances hold references to modules, so the module can only be dropped
+    // when all instances are dropped. For a given module, we can have at most one
+    // instance for each scheduling group.
+    // This function is called each time a new instance is created for a given module.
+    // If there were no instances for the module before, i.e. this module was not
+    // compiled, the module is compiled and the size of the compiled code is added
+    // to the total size of compiled code. If the total size of compiled code exceeds
+    // the maximum size as a result of this, the function will evict modules until
+    // there is enough space for the new module. If it is not possible, the function
+    // will throw an exception. If this function succeeds, the counter of instances
+    // for the module is increased by one.
+    void track_module_ref(wasmtime::Module& module, wasmtime::Engine& engine);
+
+    // This function is called each time an instance for a given module is dropped.
+    // If the counter of instances for the module reaches zero, the module is dropped
+    // and the size of the compiled code is subtracted from the total size of compiled code.
+    void remove_module_ref(wasmtime::Module& module) noexcept;
+
+    // When a WASM UDF is executed, a separate stack is first allocated for it.
+    // This stack is used by the WASM code and it is not tracked by the seastar allocator.
+    // This function will evict cached modules until the stack can be allocated. If enough
+    // memory can't be freed, the function will throw an exception.
+    void reserve_wasm_stack();
+
+    // This function should be called after a WASM UDF finishes execution. Its stack is then
+    // destroyed and this function accounts for the freed memory.
+    void free_wasm_stack() noexcept;
```
Nit: Is there a reason why those methods are public and not private? They seem to be internal to how the instance cached works. The methods `track_module_ref`/`remove_module_ref` now should probably be exclusively used by `module_handle`, so they could also be made private and `module_handle` could become a friend of `instance_cache`.

piodul reviewed (approved):
One last suggestion from my side, otherwise LGTM.

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/3710/

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/3721/

wmitros commented:
The last rebase only fixed a conflict with master.

Assuming I didn't mess up the last nitpick, we got a LGTM from @piodul, so can you take a (maybe final) look @avikivity?

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/3731/
